README
NAME
    `Net::Async::HTTP' - use HTTP with `IO::Async'

SYNOPSIS
     use IO::Async::Loop;
     use Net::Async::HTTP;
     use URI;

     my $loop = IO::Async::Loop->new();

     my $http = Net::Async::HTTP->new();

     $loop->add( $http );

     my ( $response ) = $http->do_request(
        uri => URI->new( "http://www.cpan.org/" ),
     )->get;

     print "Front page of http://www.cpan.org/ is:\n";
     print $response->as_string;

DESCRIPTION
    This object class implements an asynchronous HTTP user agent. It sends
    requests to servers, returning Future instances to yield responses when
    they are received. The object supports multiple concurrent connections
    to servers, and allows multiple requests in the pipeline to any one
    connection. Normally, only one such object will be needed per program to
    support any number of requests.

    As well as using futures the module also supports a callback-based
    interface.

    This module optionally supports SSL connections, if IO::Async::SSL is
    installed. If so, SSL can be requested either by passing a URI with the
    `https' scheme, or by passing a true value as the `SSL' parameter.

  Connection Pooling
    There are three ways in which connections to HTTP server hosts are
    managed by this object, controlled by the value of
    `max_connections_per_host'. This controls when new connections are
    established to servers, as compared to waiting for existing connections
    to be free, as new requests are made to them.

    They are:

    max_connections_per_host = 1
      This is the default setting. In this mode, there will be one
      connection per host on which there are active or pending requests. If
      new requests are made while an existing one is outstanding, they will
      be queued to wait for it.

      If pipelining is active on the connection (because both the `pipeline'
      option is true and the connection is known to be an HTTP/1.1 server),
      then requests will be pipelined into the connection awaiting their
      response. If not, they will be queued awaiting a response to the
      previous before sending the next.

    max_connections_per_host > 1
      In this mode, there can be more than one connection per host. If a new
      request is made, it will try to re-use idle connections if there are
      any, or if they are all busy it will create a new connection to the
      host, up to the configured limit.

    max_connections_per_host = 0
      In this mode, there is no upper limit to the number of connections per
      host. Every new request will try to reuse an idle connection, or else
      create a new one if all the existing ones are busy.

    These modes all apply per hostname / server port pair; they do not
    affect the behaviour of connections made to differing hostnames, or
    differing ports on the same hostname.

PARAMETERS
    The following named parameters may be passed to `new' or `configure':

  user_agent => STRING
    A string to set in the `User-Agent' HTTP header. If not supplied, one
    will be constructed that declares `Net::Async::HTTP' and the version
    number.

  max_redirects => INT
    Optional. How many levels of redirection to follow. If not supplied,
    will default to 3. Give 0 to disable redirection entirely.

  max_in_flight => INT
    Optional. The maximum number of in-flight requests to allow per host
    when pipelining is enabled and supported on that host. If more requests
    are made over this limit they will be queued internally by the object
    and not sent to the server until responses are received. If not
    supplied, will default to 4. Give 0 to disable the limit entirely.

  max_connections_per_host => INT
    Optional. Controls the maximum number of connections per hostname/server
    port pair, before requests will be queued awaiting one to be free. If
    not supplied, will default to 1. Give 0 to disable the limit entirely.
    See also the Connection Pooling section documented above.

  timeout => NUM
    Optional. How long in seconds to wait before giving up on a request. If
    not supplied then no default will be applied, and no timeout will take
    place.

  stall_timeout => NUM
    Optional. How long in seconds to wait after each write or read of data
    on a socket, before giving up on a request. This may be more useful than
    `timeout' on large-file operations, as it will not time out provided
    that regular progress is still being made.

  proxy_host => STRING
  proxy_port => INT
    Optional. Default values to apply to each `request' method.

  cookie_jar => HTTP::Cookies
    Optional. A reference to a HTTP::Cookies object. Will be used to set
    cookies in requests and store them from responses.

  pipeline => BOOL
    Optional. If false, disables HTTP/1.1-style request pipelining.

  local_host => STRING
  local_port => INT
  local_addrs => ARRAY
  local_addr => HASH or ARRAY
    Optional. Parameters to pass on to the `connect' method used to connect
    sockets to HTTP servers. Sets the local socket address to `bind()' to.
    For more detail, see the documentation in IO::Async::Connector.

  fail_on_error => BOOL
    Optional. Affects the behaviour of response handling when a `4xx' or
    `5xx' response code is received. When false, these responses will be
    processed as other responses and yielded as the result of the future, or
    passed to the `on_response' callback. When true, such an error response
    causes the future to fail, or the `on_error' callback to be invoked.

    The HTTP response and request objects will be passed as well as the code
    and message, and the failure name will be `http'.

     ( $code_message, "http", $response, $request ) = $f->failure

     $on_error->( "$code $message", $response, $request )

  read_len => INT
  write_len => INT
    Optional. Used to set the reading and writing buffer lengths on the
    underlying `IO::Async::Stream' objects that represent connections to the
    server. If not define, a default of 64 KiB will be used.

  ip_tos => INT or STRING
    Optional. Used to set the `IP_TOS' socket option on client sockets. If
    given, should either be a `IPTOS_*' constant, or one of the string names
    `lowdelay', `throughput', `reliability' or `mincost'. If undefined or
    left absent, no option will be set.

  decode_content => BOOL
    Optional. If true, incoming responses that have a recognised
    `Content-Encoding' are handled by the module, and decompressed content
    is passed to the body handling callback or returned in the
    `HTTP::Response'. See CONTENT DECODING below for details of which
    encoding methods are recognised. When this option is enabled, outgoing
    requests also have the `Accept-Encoding' header added to them if it does
    not already exist.

    Currently the default is false, because this behaviour is new, but it
    may default to true in a later version. Applications which care which
    behaviour applies should set this to a defined value to ensure it
    doesn't change.

  SSL_*
    Additionally, any parameters whose names start with `SSL_' will be
    stored and passed on requests to perform SSL requests. This simplifies
    configuration of common SSL parameters.

METHODS
    The following methods documented with a trailing call to `->get' return
    Future instances.

    When returning a Future, the following methods all indicate HTTP-level
    errors using the Future failure name of `http'. If the error relates to
    a specific response it will be included. The original request is also
    included.

     $f->fail( $message, "http", $response, $request )

  $response = $http->do_request( %args )->get
    Send an HTTP request to a server, returning a Future that will yield the
    response. The request may be represented by an HTTP::Request object, or
    a URI object, depending on the arguments passed.

    The following named arguments are used for `HTTP::Request's:

    request => HTTP::Request
            A reference to an `HTTP::Request' object

    host => STRING
            Hostname of the server to connect to

    port => INT or STRING
            Optional. Port number or service of the server to connect to. If
            not defined, will default to `http' or `https' depending on
            whether SSL is being used.

    SSL => BOOL
            Optional. If true, an SSL connection will be used.

    The following named arguments are used for `URI' requests:

    uri => URI or STRING
            A reference to a `URI' object, or a plain string giving the
            request URI. If the scheme is `https' then an SSL connection
            will be used.

    method => STRING
            Optional. The HTTP method. If missing, `GET' is used.

    content => STRING or ARRAY ref
            Optional. The body content to use for `PUT' or `POST' requests.
            If this is a plain scalar instead of an ARRAY ref, it will not
            be form encoded. In this case, a `content_type' field must also
            be supplied to describe it.

    content_type => STRING
            The type of non-form data `content'.

    user => STRING
    pass => STRING
            Optional. If both are given, the HTTP Basic Authorization header
            will be sent with these details.

    proxy_host => STRING
    proxy_port => INT
            Optional. Override the hostname or port number implied by the
            URI.

    For either request type, it takes the following arguments:

    request_body => STRING | CODE | Future
            Optional. Allows request body content to be generated by a
            future or callback, rather than being provided as part of the
            `request' object. This can either be a plain string, a `CODE'
            reference to a generator function, or a future.

            As this is passed to the underlying IO::Async::Stream `write'
            method, the usual semantics apply here. If passed a `CODE'
            reference, it will be called repeatedly whenever it's safe to
            write. The code should should return `undef' to indicate
            completion. If passed a `Future' it is expected to eventually
            yield the body value.

            As with the `content' parameter, the `content_type' field should
            be specified explicitly in the request header, as should the
            content length (typically via the HTTP::Request `content_length'
            method). See also examples/PUT.pl.

    expect_continue => BOOL
            Optional. If true, sets the `Expect' request header to the value
            `100-continue' and does not send the `request_body' parameter
            until a `100 Continue' response is received from the server. If
            an error response is received then the `request_body' code, if
            present, will not be invoked.

    on_redirect => CODE
            Optional. A callback that is invoked if a redirect response is
            received, before the new location is fetched. It will be passed
            the response and the new URL.

             $on_redirect->( $response, $location )

    on_body_write => CODE
            Optional. A callback that is invoked after each successful
            `syswrite' of the body content. This may be used to implement an
            upload progress indicator or similar. It will be passed the
            total number of bytes of body content written so far (i.e.
            excluding bytes consumed in the header).

             $on_body_write->( $written )

    max_redirects => INT
            Optional. How many levels of redirection to follow. If not
            supplied, will default to the value given in the constructor.

    timeout => NUM
    stall_timeout => NUM
            Optional. Overrides the object's configured timeout values for
            this one request. If not specified, will use the configured
            defaults.

            On a timeout, the returned future will fail with either
            `timeout' or `stall_timeout' as the operation name.

             ( $message, "timeout" ) = $f->failure

  $http->do_request( %args )

Changes
Revision history for Net-Async-HTTP

0.37    2014/12/13 15:06:06
        [CHANGES]
         * Major refactoring of Net::Async::HTTP::Connection internal closure-
           generation logic

        [BUGFIXES]
         * Cancel pending connect Futures if pipelining means an existing
           connection can be reused; handles failures better (RT99142)
         * Attempt to handle the case where $f->done itself throws an
           exception by at least maintaining internal state (RT100066)

0.36    2014/10/14 12:03:07
        [CHANGES]
         * Set the content of generated requests if a 'content' param is given
           (RT97654)
         * Store object-wide SSL params for convenience (RT98514)
         * Clarify the final return value of the on_body_chunk handler
           (RT98278)
         * Nicer error message when a non-HTTP response header is received
           (RT93231)

        [BUGFIXES]
         * 3xx responses without a Location header, and non-(GET|HEAD)
           requests should not be redirected (RT98093)

0.35    2014/03/30 20:58:34
        [BUGFIXES]
         * Reparse URI scheme after redirect, so host/port/SSLness are reset
           correctly (RT94303)
         * Requeue the next pending request after an HTTP/1.0-style EOF
           (RT94304)
         * Require at least Compress::Bzip2 2.10 to do bzip2 decompression
        
        Bugfixes sponsored by Cisco (http://www.cisco.com/)

0.34    2014/03/27 18:26:43
        [BUGFIXES]
         * Don't attempt to reconnect ready queue items that already have a
           pending connection attempt (RT92904)
         * Don't close over $responder, thus avoiding a reference cycle leak
           (RT93232 / RT92728)
         * Avoid relying on strong forward references in Future, by creating
           intentional cycles on pending Futures. Workaround for bugfix in
           upcoming Future release.

        Bugfixes sponsored by NET-A-PORTER (http://www.net-a-porter.com/)

0.33    2014/01/22 23:48:40
        [CHANGES]
         * Include connection flieno in notifier_name in case of multiple
           connections to a given host+port
         * Use different Future failure operation names for timeout and
           stall_timeout

        [BUGFIXES]
         * Zlib compression needs Compress::Raw::Zlib 2.057 or newer
         * Ensure that unpipelined requests can be cancelled
         * Ensure that other pending requests on a connection are also
           errored out when connection receives EOF

0.32    2013/11/19 14:09:18
        [CHANGES]
         * Optionally decode encoded content and set the Accept-Encoding
           header of outbound requests. Currently defaults off, but may become
           true in a later version.
        
        [BUGFIXES]
         * Handle resolve/connect errors and pipeline queue flushing after
           cancellation correctly

0.31    2013/11/04 18:04:30
        [CHANGES]
         * Allow a plain string URI, upgrading it to a URI object (RT89269)
         * Added ->POST shortcut method, similar to ->GET and ->HEAD
         * Extract HTTP basic auth. information from the URL in an
           HTTP::Request (RT89775)

        [BUGFIXES]
         * Remember to keep resetting the stall timer when receiving body
           content
         * Fix awkward race between stall_timer and ->cancel
         * Note dependency on Test::Refcount

0.30    2013/10/20 02:08:01
        [CHANGES]
         * Use Future->new->fail to return failing immediate Futures
         * Ensure Future failures all use standard IO::Async style of
           ($message, NAME, ...)

        [BUGFIXES]
         * Ensure that fail_on_error doesn't cause non-error responses to fail
           (RT88996)

0.29    2013/09/22 02:50:39
        [CHANGES]
         * Added 'on_body_write' callback
         * Reordering of documentation to emphasise futures first

        [BUGFIXES]
         * Fix handling of chunk header when split across reads

0.28    2013/09/13 20:40:14
        [CHANGES]
         * Rewrite back into being a subclass of IO::Async::Stream instead of
           using IO::Async::Protocol::Stream
         * IO::Async::Stream 0.59 allows stall timeout detection during writes
           now as well

        [BUGFIXES]
         * Closed connections should drop the ready queue, so they don't
           accidentally try to process a second request
         * Failure handling of corrupted chunk headers during chunked transfer
           encoding

0.27    BUGFIXES:
         * Ensure that $f->cancel on a request Future actually works properly

0.26    CHANGES:
         * Added 'stall_timeout', at least for read operations

        BUGFIXES:
         * Fix NaHTTP+NaHTTP::Server unit tests for minimum version
           requirements

0.25    BUGFIXES:
         * Defend against ->setsockopt detecting SvPOK() of string values
         * Test for both "Connection closed" and "Connection closed while
           awaiting header" as OSes differ in behaviour

0.24    CHANGES:
         * Optionally set IP_TOS field in sockets
         * Remember to pass 'SSL' parameter around for requests on 'https'
           scheme

0.23    BUGFIXES:
         * Fix t/14conn-max.t failures - disable pipelining, wait for correct
           peersock connection to become ready (it may not be [0])

0.22    CHANGES:
         * Allow more control over per-host concurrency; either a pool with a
           limit, or boundless growth of concurrency

0.21    CHANGES:
         * Optional mode in which 4xx and 5xx HTTP errors are turned into
           callback errors or Future failures
         * Handle incoming 1xx informational responses
         * Optionally set 'Expect' header for 100 Continue response

0.20    CHANGES:
         * Improvements to timeout and cancellation handling
         * Updated for Future 0.12; pass 'return' argument to
           Future::Utils::repeat
         * Use newer IO::Async::OS->socketpair in unit tests

0.19    CHANGES:

lib/Net/Async/HTTP/Connection.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2008-2014 -- leonerd@leonerd.org.uk

package Net::Async::HTTP::Connection;

use strict;
use warnings;

our $VERSION = '0.37';

use Carp;

use base qw( IO::Async::Stream );
IO::Async::Stream->VERSION( '0.59' ); # ->write( ..., on_write )

use Net::Async::HTTP::StallTimer;

use HTTP::Response;

my $CRLF = "\x0d\x0a"; # More portable than \r\n

use Struct::Dumb;
struct Responder => [qw( on_read on_error stall_timer is_done )];

# Detect whether HTTP::Message properly trims whitespace in header values. If
# it doesn't, we have to deploy a workaround to fix them up.
#   https://rt.cpan.org/Ticket/Display.html?id=75224
use constant HTTP_MESSAGE_TRIMS_LWS => HTTP::Message->parse( "Name:   value  " )->header("Name") eq "value";

=head1 NAME

C<Net::Async::HTTP::Connection> - HTTP client protocol handler

=head1 DESCRIPTION

This class provides a connection to a single HTTP server, and is used
internally by L<Net::Async::HTTP>. It is not intended for general use.

=cut

sub _init
{
   my $self = shift;
   $self->SUPER::_init( @_ );

   $self->{requests_in_flight} = 0;
}

sub configure
{
   my $self = shift;
   my %params = @_;

   foreach (qw( pipeline max_in_flight ready_queue decode_content )) {
      $self->{$_} = delete $params{$_} if exists $params{$_};
   }

   if( my $on_closed = $params{on_closed} ) {
      $params{on_closed} = sub {
         my $self = shift;

         $self->debug_printf( "CLOSED" );

         $self->error_all( "Connection closed" );

         undef $self->{ready_queue};
         $on_closed->( $self );
      };
   }

   croak "max_in_flight parameter required, may be zero" unless defined $self->{max_in_flight};

   $self->SUPER::configure( %params );
}

sub should_pipeline
{
   my $self = shift;
   return $self->{pipeline} &&
          $self->{can_pipeline} &&
          ( !$self->{max_in_flight} || $self->{requests_in_flight} < $self->{max_in_flight} );
}

sub connect
{
   my $self = shift;
   my %args = @_;

   $self->debug_printf( "CONNECT $args{host}:$args{service}" );

   defined wantarray or die "VOID ->connect";

   $self->SUPER::connect(
      socktype => "stream",
      %args
   )->on_done( sub {
      $self->debug_printf( "CONNECTED" );
      $self->ready;
   });
}

sub ready
{
   my $self = shift;

   my $queue = $self->{ready_queue} or return;

   if( $self->should_pipeline ) {
      $self->debug_printf( "READY pipelined" );
      while( @$queue && $self->should_pipeline ) {
         my $ready = shift @$queue;
         my $f = $ready->future;
         next if $f->is_cancelled;

         $ready->connecting and $ready->connecting->cancel;

         $f->done( $self );
      }
   }
   elsif( @$queue and $self->is_idle ) {
      $self->debug_printf( "READY non-pipelined" );
      while( @$queue ) {
         my $ready = shift @$queue;
         my $f = $ready->future;
         next if $f->is_cancelled;

         $ready->connecting and $ready->connecting->cancel;

         $f->done( $self );
         last;
      }
   }
   else {
      $self->debug_printf( "READY cannot-run queue=%d idle=%s",
         scalar @$queue, $self->is_idle ? "Y" : "N");
   }
}

sub is_idle
{
   my $self = shift;
   return $self->{requests_in_flight} == 0;
}

sub on_read
{
   my $self = shift;
   my ( $buffref, $closed ) = @_;

   while( my $head = $self->{responder_queue}[0]) {
      shift @{ $self->{responder_queue} } and next if $head->is_done;

      $head->stall_timer->reset if $head->stall_timer;

      my $ret = $head->on_read->( $self, $buffref, $closed, $head );

      if( defined $ret ) {
         return $ret if !ref $ret;

         $head->on_read = $ret;
         return 1;
      }

      $head->is_done or die "ARGH: undef return without being marked done";

      shift @{ $self->{responder_queue} };
      return 1 if !$closed and length $$buffref;
      return;
   }

   # Reinvoked after switch back to baseline, but may be idle again
   return if $closed or !length $$buffref;

   $self->invoke_error( "Spurious on_read of connection while idle",
      http_connection => read => $$buffref );
   $$buffref = "";
}

sub on_write_eof
{
   my $self = shift;
   $self->error_all( "Connection closed", http => undef, undef );
}

sub error_all
{
   my $self = shift;

   while( my $head = shift @{ $self->{responder_queue} } ) {
      $head->on_error->( @_ ) unless $head->is_done;
   }
}

sub request
{
   my $self = shift;
   my %args = @_;

   my $on_header = $args{on_header} or croak "Expected 'on_header' as a CODE ref";

   my $req = $args{request};
   ref $req and $req->isa( "HTTP::Request" ) or croak "Expected 'request' as a HTTP::Request reference";

   $self->debug_printf( "REQUEST %s %s", $req->method, $req->uri );

   my $request_body = $args{request_body};
   my $expect_continue = !!$args{expect_continue};

   my $method = $req->method;

   if( $method eq "POST" or $method eq "PUT" or length $req->content ) {
      $req->init_header( "Content-Length", length $req->content );
   }

   if( $expect_continue ) {
      $req->init_header( "Expect", "100-continue" );
   }

   if( $self->{decode_content} ) {
      #$req->init_header( "Accept-Encoding", Net::Async::HTTP->can_decode )
      $req->init_header( "Accept-Encoding", "gzip" );
   }

   my $f = $self->loop->new_future
      ->set_label( "$method " . $req->uri );

   # TODO: Cancelling a request Future shouldn't necessarily close the socket
   # if we haven't even started writing the request yet. But we can't know
   # that currently.
   $f->on_cancel( sub { $self->close_now } );

   my $stall_timer;
   if( $args{stall_timeout} ) {
      $stall_timer = Net::Async::HTTP::StallTimer->new(
         delay => $args{stall_timeout},
         future => $f,
      );
      $self->add_child( $stall_timer );
      # Don't start it yet

      my $remove_timer = sub {
         $self->remove_child( $stall_timer ) if $stall_timer;
         undef $stall_timer;
      };

      $f->on_ready( $remove_timer );
      $f->on_cancel( $remove_timer );
   }

   my $on_body_write;
   if( $stall_timer or $args{on_body_write} ) {
      my $inner_on_body_write = $args{on_body_write};
      my $written = 0;
      $on_body_write = sub {
         $stall_timer->reset if $stall_timer;
         $inner_on_body_write->( $written += $_[1] ) if $inner_on_body_write;
      };
   }

   my $write_request_body = defined $request_body ? sub {
      my ( $self ) = @_;
      $self->write( $request_body,
         on_write => $on_body_write
      );
   } : undef;

   # Unless the request method is CONNECT, the URL is not allowed to contain
   # an authority; only path
   # Take a copy of the headers since we'll be hacking them up
   my $headers = $req->headers->clone;
   my $path;
   if( $method eq "CONNECT" ) {
      $path = $req->uri->as_string;
   }
   else {
      my $uri = $req->uri;
      $path = $uri->path_query;
      $path = "/$path" unless $path =~ m{^/};
      my $authority = $uri->authority;
      if( defined $authority and
          my ( $user, $pass, $host ) = $authority =~ m/^(.*?):(.*)@(.*)$/ ) {
         $headers->init_header( Host => $host );
         $headers->authorization_basic( $user, $pass );
      }
      else {
         $headers->init_header( Host => $authority );
      }
   }

   my $protocol = $req->protocol || "HTTP/1.1";
   my @headers = ( "$method $path $protocol" );
   $headers->scan( sub { push @headers, "$_[0]: $_[1]" } );

   $stall_timer->start if $stall_timer;
   $stall_timer->reason = "writing request" if $stall_timer;

   my $on_header_write = $stall_timer ? sub { $stall_timer->reset } : undef;

   $self->write( join( $CRLF, @headers ) .
                 $CRLF . $CRLF,
                 on_write => $on_header_write );

   $self->write( $req->content,
                 on_write => $on_body_write ) if length $req->content;
   $write_request_body->( $self ) if $write_request_body and !$expect_continue;

   $self->write( "", on_flush => sub {
      return unless $stall_timer; # test again in case it was cancelled in the meantime
      $stall_timer->reset;
      $stall_timer->reason = "waiting for response";
   }) if $stall_timer;

   $self->{requests_in_flight}++;

   push @{ $self->{responder_queue} }, Responder(
      $self->_mk_on_read_header(
         $req, $args{previous_response}, $expect_continue ? $write_request_body : undef, $on_header, $stall_timer, $f
      ),
      sub { $f->fail( @_ ) unless $f->is_ready; }, # on_error
      $stall_timer,
      0, # is_done
   );

   return $f;
}

sub _mk_on_read_header
{
   shift; # $self
   my ( $req, $previous_response, $write_request_body, $on_header, $stall_timer, $f ) = @_;

   sub {
      my ( $self, $buffref, $closed, $responder ) = @_;

      if( $stall_timer ) {
         $stall_timer->reason = "receiving response header";
         $stall_timer->reset;
      }

      if( length $$buffref >= 4 and $$buffref !~ m/^HTTP/ ) {
         $self->debug_printf( "ERROR fail" );
         $f->fail( "Did no receive HTTP response from server", http => undef, $req ) unless $f->is_cancelled;
         $self->close_now;
      }

      unless( $$buffref =~ s/^(.*?$CRLF$CRLF)//s ) {
         if( $closed ) {
            $self->debug_printf( "ERROR closed" );
            $f->fail( "Connection closed while awaiting header", http => undef, $req ) unless $f->is_cancelled;
            $self->close_now;
         }
         return 0;
      }

      my $header = HTTP::Response->parse( $1 );
      # HTTP::Response doesn't strip the \rs from this
      ( my $status_line = $header->status_line ) =~ s/\r$//;

      unless( HTTP_MESSAGE_TRIMS_LWS ) {
         my @headers;
         $header->scan( sub {
            my ( $name, $value ) = @_;
            s/^\s+//, s/\s+$// for $value;
            push @headers, $name => $value;
         } );
         $header->header( @headers ) if @headers;
      }

      my $protocol = $header->protocol;
      if( $protocol =~ m{^HTTP/1\.(\d+)$} and $1 >= 1 ) {
         $self->{can_pipeline} = 1;
      }

      if( $header->code =~ m/^1/ ) { # 1xx is not a final response
         $self->debug_printf( "HEADER [provisional] %s", $status_line );
         $write_request_body->( $self ) if $write_request_body;
         return 1;
      }

      $header->request( $req );
      $header->previous( $previous_response ) if $previous_response;

      $self->debug_printf( "HEADER %s", $status_line );

      my $on_body_chunk = $on_header->( $header );

      my $code = $header->code;

      my $content_encoding = $header->header( "Content-Encoding" );

      my $decoder;
      if( $content_encoding and
          $decoder = Net::Async::HTTP->can_decode( $content_encoding ) ) {
         $header->init_header( "X-Original-Content-Encoding" => $header->remove_header( "Content-Encoding" ) );
      }

      # can_pipeline is set for HTTP/1.1 or above; presume it can keep-alive if set
      my $connection_close = lc( $header->header( "Connection" ) || ( $self->{can_pipeline} ? "keep-alive" : "close" ) )
                              eq "close";

      if( $connection_close ) {
         $self->{max_in_flight} = 1;
      }
      elsif( defined( my $keep_alive = lc( $header->header("Keep-Alive") || "" ) ) ) {
         my ( $max ) = ( $keep_alive =~ m/max=(\d+)/ );
         $self->{max_in_flight} = $max if $max && $max < $self->{max_in_flight};
      }

      my $on_more = sub {
         my ( $chunk ) = @_;

         if( $decoder and not eval { $chunk = $decoder->( $chunk ); 1 } ) {
            $self->debug_printf( "ERROR decode failed" );
            $f->fail( "Decode error $@", http => undef, $req );
            $self->close;
            return undef;
         }

         $on_body_chunk->( $chunk );

         return 1;
      };
      my $on_done = sub {
         # TODO: IO::Async probably ought to do this. We need to fire the
         # on_closed event _before_ calling on_body_chunk, to clear the
         # connection cache in case another request comes - e.g. HEAD->GET
         $self->close if $connection_close;

         my $final;
         if( $decoder and not eval { $final = $decoder->(); 1 } ) {
            $self->debug_printf( "ERROR decode failed" );
            $f->fail( "Decode error $@", http => undef, $req );
            $self->close;
            return undef;
         }

         $on_body_chunk->( $final ) if length $final;

         my $response = $on_body_chunk->();
         my $e = eval { $f->done( $response ) unless $f->is_cancelled; 1 } ? undef : $@;

         $self->{requests_in_flight}--;
         $self->debug_printf( "DONE remaining in-flight=$self->{requests_in_flight}" );
         $self->ready;

         if( defined $e ) {
            chomp $e;
            $self->invoke_error( $e, perl => );
            # This might not return, if it top-level croaks
         }

         return undef; # Finished
      };

      # RFC 2616 says "HEAD" does not have a body, nor do any 1xx codes, nor
      # 204 (No Content) nor 304 (Not Modified)
      if( $req->method eq "HEAD" or $code =~ m/^1..$/ or $code eq "204" or $code eq "304" ) {
         $self->debug_printf( "BODY done [none]" );
         $responder->is_done++;

         return $on_done->();
      }

      my $transfer_encoding = $header->header( "Transfer-Encoding" );
      my $content_length    = $header->content_length;

      if( defined $transfer_encoding and $transfer_encoding eq "chunked" ) {
         $self->debug_printf( "BODY chunks" );

         $stall_timer->reason = "receiving body chunks" if $stall_timer;
         return $self->_mk_on_read_chunked( $req, $on_more, $on_done, $f );
      }
      elsif( defined $content_length ) {
         $self->debug_printf( "BODY length $content_length" );

         if( $content_length == 0 ) {
            $self->debug_printf( "BODY done [length=0]" );
            $responder->is_done++;

            return $on_done->();
         }

         $stall_timer->reason = "receiving body" if $stall_timer;
         return $self->_mk_on_read_length( $content_length, $req, $on_more, $on_done, $f );
      }
      else {
         $self->debug_printf( "BODY until EOF" );

         $stall_timer->reason = "receiving body until EOF" if $stall_timer;
         return $self->_mk_on_read_until_eof( $req, $on_more, $on_done, $f );
      }
   };
}

sub _mk_on_read_chunked
{
   shift; # $self
   my ( $req, $on_more, $on_done, $f ) = @_;

   my $chunk_length;

   sub {
      my ( $self, $buffref, $closed, $responder ) = @_;

      if( !defined $chunk_length and $$buffref =~ s/^(.*?)$CRLF// ) {
         my $header = $1;

         # Chunk header
         unless( $header =~ s/^([A-Fa-f0-9]+).*// ) {
            $f->fail( "Corrupted chunk header", http => undef, $req ) unless $f->is_cancelled;
            $self->close_now;
            return 0;
         }

         $chunk_length = hex( $1 );
         return 1 if $chunk_length;

         return $self->_mk_on_read_chunk_trailer( $req, $on_more, $on_done, $f );
      }

      # Chunk is followed by a CRLF, which isn't counted in the length;
      if( defined $chunk_length and length( $$buffref ) >= $chunk_length + 2 ) {
         # Chunk body
         my $chunk = substr( $$buffref, 0, $chunk_length, "" );
         undef $chunk_length;

         unless( $$buffref =~ s/^$CRLF// ) {
            $self->debug_printf( "ERROR chunk without CRLF" );
            $f->fail( "Chunk of size $chunk_length wasn't followed by CRLF", http => undef, $req ) unless $f->is_cancelled;
            $self->close;
         }

         return $on_more->( $chunk );
      }

      if( $closed ) {
         $self->debug_printf( "ERROR closed" );
         $f->fail( "Connection closed while awaiting chunk", http => undef, $req ) unless $f->is_cancelled;
      }
      return 0;
   };
}

sub _mk_on_read_chunk_trailer
{
   shift; # $self
   my ( $req, $on_more, $on_done, $f ) = @_;

   my $trailer = "";

   sub {
      my ( $self, $buffref, $closed, $responder ) = @_;

      if( $closed ) {
         $self->debug_printf( "ERROR closed" );
         $f->fail( "Connection closed while awaiting chunk trailer", http => undef, $req ) unless $f->is_cancelled;
      }

      $$buffref =~ s/^(.*)$CRLF// or return 0;
      $trailer .= $1;

      return 1 if length $1;

      # TODO: Actually use the trailer

      $self->debug_printf( "BODY done [chunked]" );
      $responder->is_done++;

      return $on_done->();
   };
}

sub _mk_on_read_length
{
   shift; # $self
   my ( $content_length, $req, $on_more, $on_done, $f ) = @_;

   sub {
      my ( $self, $buffref, $closed, $responder ) = @_;

      # This will truncate it if the server provided too much
      my $content = substr( $$buffref, 0, $content_length, "" );
      $content_length -= length $content;

      return undef unless $on_more->( $content );

      if( $content_length == 0 ) {
         $self->debug_printf( "BODY done [length]" );
         $responder->is_done++;

         return $on_done->();
      }

      if( $closed ) {
         $self->debug_printf( "ERROR closed" );
         $f->fail( "Connection closed while awaiting body", http => undef, $req ) unless $f->is_cancelled;
      }
      return 0;
   };
}

sub _mk_on_read_until_eof
{
   shift; # $self
   my ( $req, $on_more, $on_done, $f ) = @_;

   sub {
      my ( $self, $buffref, $closed, $responder ) = @_;

      my $content = $$buffref;
      $$buffref = "";

      return undef unless $on_more->( $content );

      return 0 unless $closed;

      $self->debug_printf( "BODY done [eof]" );
      $responder->is_done++;
      return $on_done->();
   };
}

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;
